### PR TITLE
track slack link clicks

### PR DESCRIPTION
**Pull request by camhashemi**
@ryanpbrewster @leogrim 

edit: updated. now using this task https://app.asana.com/0/32728208409723/98135038182940

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala [hunk 1]
```diff
@@ -14,11 +14,22 @@ import play.utils.UriEncoding
 
 import scala.util.Try
 
-case class RedirectTrackingParameters( // can be augmented, fields can be made optional as needed
-  eventType: EventType,
-  action: String,
-  slackUserId: SlackUserId,
-  slackTeamId: SlackTeamId)
+case class RedirectTrackingParameters(data: Map[String, String]) {
+  def toQueryParams = data.map { case (k, v) => s"$k=$v" }.mkString("&")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala]: Be sure to URL-escape these

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala [hunk 2]
```diff
@@ -14,11 +14,22 @@ import play.utils.UriEncoding
...
+  def toHeimdalContext = HeimdalContext(data.map { case (k, v) => k -> ContextStringData(v) })
+}
+object RedirectTrackingParameters {
+  def apply(params: (String, String)*): RedirectTrackingParameters = RedirectTrackingParameters(params.toMap)
+  def fromQueryParams(paramString: String): RedirectTrackingParameters = {
+    val justParams = if (paramString.headOption.contains('?')) paramString.drop(1) else paramString
+    val data = justParams.split("&").foldLeft(Map.empty[String, String]) {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala]: `split` should take a char here. If you pass a string it is interpreted as a regex.

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala [hunk 3]
```diff
@@ -14,11 +14,22 @@ import play.utils.UriEncoding
...
+      case (map, kv) =>
+        val Array(key, value) = kv.split("=").take(2)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala]: same thing here

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala [hunk 4]
```diff
@@ -34,30 +45,16 @@ class KifiUrlRedirectHelper @Inject() (app: Application) {
...
-    val keyValueString = s"eventType=${params.eventType.name}&action=${params.action}&slackUserId=${params.slackUserId.value}&slackTeamId=${params.slackTeamId.value}"
-    val signedTrackingParams = cipher.encrypt(keyValueString)
-    signedTrackingParams
-  }
+  def signTrackingParams(params: RedirectTrackingParameters): String = cipher.encrypt(params.toQueryParams)
 
   def parseKifiUrlRedirect(signedUrl: String, signedTrackingParams: Option[String]): Option[(String, Option[RedirectTrackingParameters])] = {
     val urlOpt = Crypto.extractSignedToken(URLDecoder.decode(signedUrl, "ascii"))
-    val trackingParams = signedTrackingParams.flatMap(params => extractTrackingParams(URLDecoder.decode(params, "ascii")))
+    val trackingParams = signedTrackingParams.map(params => extractTrackingParams(URLDecoder.decode(params, "ascii")))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala:24]: I'm not sure whether `"ascii"` or `"UTF-8"` is right. For URLs I think ascii is good, but I ran into issues with it while interacting with Slack. Good luck.

File: kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala [hunk 5]
```diff
@@ -90,6 +90,16 @@ object NotificationCategory {
     val ORGANIZATION_INVITATION = NotificationCategory("visitor_organization_invitation")
     val BILLING = NotificationCategory("billing")
 
+    val INTEGRATION_WELCOME = NotificationCategory("integration_welcome")
+    val INTEGRATOR_PRESYNC = NotificationCategory("integrator_presync")
+    val INTEGRATOR_POSTSYNC = NotificationCategory("integrator_postsync")
+    val TEAM_DIGEST = NotificationCategory("team_digest")
+    val PERSONAL_DIGEST = NotificationCategory("personal_digest")
+    val LIBRARY_DIGEST = NotificationCategory("library_digest")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala:98]: If we're talking Slack here, we don't have a library or channel digest anymore.

File: kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala [hunk 6]
```diff
@@ -90,6 +90,16 @@ object NotificationCategory {
...
+    val NEW_KEEP = NotificationCategory("new_keep")
+    val MIRRORED_COMMENT = NotificationCategory("mirrored_comment")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala]: want to call this `NEW_COMMENT` for symmetry with `NEW_KEEP`?

File: kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala (hunk 5 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala:98]: these are new names, we can refine them. but the library_digest category is for the message saying "LibraryX has 10 new keeps and 8 new comments since April"

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala [hunk 7]
```diff
@@ -94,15 +94,17 @@ class PathCommander @Inject() (
...
    * MISCELLANEOUS
    */
-  def browserExtensionViaSlack(slackTeamId: SlackTeamId): Path = Path(s"s/${slackTeamId.value}/i")
+  def browserExtensionViaSlack(slackTeamId: SlackTeamId, trackingParams: Option[String]): Path = Path(s"s/${slackTeamId.value}/i") + "?" + trackingParams.getOrElse("")
   def tagSearchPath(tag: String) = PathCommander.tagSearchPath(tag)
-  def slackPersonalDigestToggle(slackTeamId: SlackTeamId, slackUserId: SlackUserId, turnOn: Boolean) =
+  def slackPersonalDigestToggle(slackTeamId: SlackTeamId, slackUserId: SlackUserId, turnOn: Boolean, trackingParams: Option[String]) =
     Path(s"s/pd/${slackTeamId.value}/${slackUserId.value}/${SlackTeamMembership.encodeTeamAndUser(slackTeamId, slackUserId)}?turnOn=$turnOn")
   def welcomePageViaSlack(basicUser: BasicUser, slackTeamId: SlackTeamId): Path = Path(s"s/${slackTeamId.value}/welcome/${basicUser.externalId.id}")
   def ownKeepsFeedPage: Path = Path(s"/?filter=own")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:102]: :(

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:102]: Don't bring that only slightly related negativity into this PR, it needs all the support it can get.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala [hunk 8]
```diff
@@ -37,13 +38,18 @@ class SlackAuthRouter @Inject() (
...
+  val heimdalContextBuilder: HeimdalContextBuilderFactory,
   val airbrake: AirbrakeNotifier,
   val userActionsHelper: UserActionsHelper,
   implicit val defaultContext: ExecutionContext,
   implicit val publicIdConfiguration: PublicIdConfiguration)
     extends UserActions with ShoeboxServiceController {
 
-  def fromSlackToInstallPage(slackTeamId: SlackTeamId) = MaybeUserAction { implicit request =>
+  def fromSlackToInstallPage(slackTeamId: SlackTeamId, signedTrackingParams: Option[String]) = MaybeUserAction { implicit request =>
+    t.map(t => trackClick(slackTeamId, t, "extension"))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala]: How do you feel about having a `MaybeUserActionWithTracking` or something that looks at `request.queryString` to do tracking?

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala (hunk 4 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala:24]: this works, we're using this for tracking slack search result clicks (separate from this PR, but re-using the helper class)

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala (hunk 8 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala]: Or go with something like the `LibraryAccessActions` and have an `MaybeUserRequest andThen Tracking`

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala]: that sounds awesome

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala (hunk 7 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:102]: How about a method on `Path`, `.withTrackingParams`? Then this can be done at the call-site.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala [hunk 9]
```diff
@@ -394,16 +418,16 @@ class SlackPushingActor @Inject() (
...
         })).withFullMarkdown.withColor(LibraryColor.BLUE.hex) +: textAndLookHeres.collect {
           case Right(Success((pointer, ref))) =>
             imageUrlRegex.findFirstIn(ref) match {
               case Some(url) =>
-                SlackAttachment.simple(DescriptionElements(SlackEmoji.magnifyingGlass, pointer --> keepLink)).withImageUrl(url)
+                SlackAttachment.simple(DescriptionElements(SlackEmoji.magnifyingGlass, pointer --> keepLink("lookHereImage"))).withImageUrl(url)
               case None =>
                 SlackAttachment.simple(DescriptionElements(
-                  SlackEmoji.magnifyingGlass, pointer --> keepLink, ": ",
+                  SlackEmoji.magnifyingGlass, pointer --> keepLink("lookHere"), ": ",
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala:435]: Do we want to have a namespace with all of these defined as `val` somewhere? Seems error-prone to manually type it out every time.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackTeamCommander.scala [hunk 10]
```diff
@@ -344,8 +345,9 @@ class SlackTeamCommanderImpl @Inject() (
     }
     import DescriptionElements._
     membershipOpt.fold(Future.failed[Unit](SlackFail.NoSuchMembership(slackTeamId, slackUserId))) { membership =>
-      val toggleLink = LinkElement(pathCommander.slackPersonalDigestToggle(slackTeamId, slackUserId, turnOn = !turnOn))
-      slackClient.sendToSlackHoweverPossible(membership.slackTeamId, membership.slackUserId.asChannel, SlackMessageRequest.fromKifi(DescriptionElements.formatForSlack(
+      val channelId = membership.slackUserId.asChannel
+      val toggleLink = LinkElement(pathCommander.slackPersonalDigestToggle(slackTeamId, slackUserId, turnOn = !turnOn, Some(slackAnalytics.generateTrackingParams(channelId, NotificationCategory.NonUser.SETTINGS_TOGGLE))))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackTeamCommander.scala]: Why do we need to annotate the link with the type of notification? Can't we infer it?

File: kifi-backend/modules/shoebox/conf/site.routes [hunk 11]
```diff
@@ -34,15 +34,15 @@ GET     /url                       @com.keepit.controllers.routing.KifiSiteRoute
...
+GET     /s/:teamId/oi/:orgId                @com.keepit.controllers.routing.SlackAuthRouter.fromSlackToOrgIntegrations(teamId: SlackTeamId, orgId: PublicId[Organization], t: Option[String] ?= None)
+GET     /s/:teamId/l/:libId                 @com.keepit.controllers.routing.SlackAuthRouter.fromSlackToLibrary(teamId: SlackTeamId, libId: PublicId[Library], t: Option[String] ?= None)
+GET     /s/:teamId/k/:keepId/:urlHash/kifi  @com.keepit.controllers.routing.SlackAuthRouter.fromSlackToKeep(teamId: SlackTeamId, keepId: PublicId[Keep], urlHash: UrlHash, onKifi: Boolean = true, t: Option[String] ?= None)
+GET     /s/:teamId/k/:keepId/:urlHash/web   @com.keepit.controllers.routing.SlackAuthRouter.fromSlackToKeep(teamId: SlackTeamId, keepId: PublicId[Keep], urlHash: UrlHash, onKifi: Boolean = false, t: Option[String] ?= None)
+GET     /s/pd/:teamId/:userId/:hash         @com.keepit.controllers.routing.SlackAuthRouter.togglePersonalDigest(teamId: SlackTeamId, userId: SlackUserId, hash: String, turnOn: Boolean, t: Option[String] ?= None)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/conf/site.routes]: One advantage of the `TrackingRequest` that looks at `request.queryString` is that we can keep these routes cleaner.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackTeamCommander.scala (hunk 10 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackTeamCommander.scala]: you can link to a user, library, keep from multiple types of notifications.

File: kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala (hunk 5 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/NotificationCategory.scala:98]: Oh :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala (hunk 9 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala:435]: yeah, all of our tracking could use a lot of that. I can do it for this small scope as a start.

File: kifi-backend/modules/shoebox/conf/site.routes (hunk 11 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/conf/site.routes]: that'd be really nice

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackAnalytics.scala [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    val contextBuilder = heimdalContextBuilder()
+    contextBuilder += ("action", "click")
+    contextBuilder += ("subaction", subaction)
+    trackSlackNotificationEvent(slackTeamId, slackChannelId, category)
+  }
+
+  def generateTrackingParams(slackChannelId: SlackChannelId, category: NotificationCategory, subaction: Option[String] = None): String = {
+    val params = Seq("slackChannelId" -> slackChannelId.value, "category" -> category.category) ++ subaction.map(s => "subaction" -> s).toSeq
+    s"t=${redirectHelper.signTrackingParams(RedirectTrackingParameters(params.toMap))}"
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackAnalytics.scala]: Can you put this method on the companion object to this class, to make it clear that it is a "pure" function and cannot open a db session?

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackAnalytics.scala]: Alternately, create a separate object to hold this method, say `SlackTrackingGenerator`

File: kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/KifiUrlRedirectHelper.scala]: You may want to look at `Query` and `Param` in `com.keepit.common.net`, which I think helps with  some of the headaches of

**Comment by camhashemi** [comment]
updated per @ryanpbrewster 's great suggestions

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala [hunk 13]
```diff
@@ -1,37 +1,40 @@
...
 
   def isEncoded: Boolean = false
 
   def absolute: String = {
     if (relative.startsWith("/")) {
-      Path.base + relative.drop(1)
+      Path.base + relative.drop(1) + (if (query != Query.empty) s"?$query" else "")
     } else {
-      Path.base + relative
+      Path.base + relative + (if (query != Query.empty) s"?$query" else "")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: Can be shortened by using `relative.stripPrefix("/")`

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: Might also suggest being defensive and assume that existing code abuses `relative` by sticking query params in there. You may want to sanitize paths when they are created by splitting out any query params hiding in the `relative` part.

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala [hunk 14]
```diff
@@ -1,37 +1,40 @@
...
     }
   }
 
-  def relative: String = value
+  def relative: String = value + (if (query != Query.empty) s"?$query" else "")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: Won't this put `?$query` in twice when you use `Path.base + relative + s"?$query"`?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala [hunk 15]
```diff
@@ -96,14 +90,14 @@ class PathCommander @Inject() (
...
+  def userPageViaSlack(basicUser: BasicUser, slackTeamId: SlackTeamId): Path = Path(s"s/${slackTeamId.value}/u/${basicUser.externalId.id}")
 
   /**
    * MISCELLANEOUS
    */
   def browserExtensionViaSlack(slackTeamId: SlackTeamId): Path = Path(s"s/${slackTeamId.value}/i")
   def tagSearchPath(tag: String) = PathCommander.tagSearchPath(tag)
-  def slackPersonalDigestToggle(slackTeamId: SlackTeamId, slackUserId: SlackUserId, turnOn: Boolean) =
-    Path(s"s/pd/${slackTeamId.value}/${slackUserId.value}/${SlackTeamMembership.encodeTeamAndUser(slackTeamId, slackUserId)}?turnOn=$turnOn")
+  def slackPersonalDigestToggle(slackTeamId: SlackTeamId, slackUserId: SlackUserId, turnOn: Boolean) = Path(s"s/pd/${slackTeamId.value}/${slackUserId.value}/${SlackTeamMembership.encodeTeamAndUser(slackTeamId, slackUserId)}").withQuery(Query("turnOn" -> turnOn.toString))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:100]: :boom: that is sweet

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:100]: I think that there is an `implicit def` from `Boolean` to `Param` somewhere, if you want to avoid the `.toString`

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala [hunk 16]
```diff
@@ -145,7 +150,7 @@ class SlackAuthRouter @Inject() (
     }
   }
 
-  def togglePersonalDigest(slackTeamId: SlackTeamId, slackUserId: SlackUserId, hash: String, turnOn: Boolean) = MaybeUserAction { implicit request =>
+  def togglePersonalDigest(slackTeamId: SlackTeamId, slackUserId: SlackUserId, hash: String, turnOn: Boolean) = (MaybeUserAction andThen SlackClickTracking(slackTeamId, "toggleDigest")) { implicit request =>
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/routing/SlackAuthRouter.scala:153]: That is so rad

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackOnboarder.scala [hunk 17]
```diff
@@ -151,8 +163,13 @@ class SlackOnboarderImpl @Inject() (
 
   private def explicitIngestionMessage(sctl: SlackChannelToLibrary, owner: BasicUser, lib: Library, slackTeam: SlackTeam)(implicit session: RSession): Option[SlackMessageRequest] = {
     import DescriptionElements._
+
+    val trackingParams = slackAnalytics.generateTrackingParams(sctl.slackChannelId, NotificationCategory.NonUser.INTEGRATION_WELCOME)
+    val welcomeLink = LinkElement(pathCommander.welcomePageViaSlack(owner, sctl.slackTeamId).withQuery(trackingParams))
+    val libraryLink = LinkElement(pathCommander.libraryPageViaSlack(lib, slackTeam.slackTeamId).withQuery(trackingParams))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackOnboarder.scala:169]: Might be cool to have this use an `implicit trackingParams: Query`. I think there is a cool way to get `implicit trackingParams: Option[Query]` that is `None` if no implicit is defined.

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala [hunk 18]
```diff
@@ -1,37 +1,40 @@
...
+sealed class Path(private val value: String, private val query: Query) {
+  def encode: EncodedPath = new EncodedPath(value, query)
   def decode: Path = this
 
   def isEncoded: Boolean = false
 
   def absolute: String = {
     if (relative.startsWith("/")) {
-      Path.base + relative.drop(1)
+      Path.base + relative.drop(1) + (if (query != Query.empty) s"?$query" else "")
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: Seems like you have this a lot. Alternate syntax ideas: `query.map("?" + _)`, or `query.urlFormat`

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala (hunk 15 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PathCommander.scala:100]: it's probably less characters than however I'd get that in the scope hehe

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala (hunk 14 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: doh, thank you

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala (hunk 13 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: :+1: `relative.stripPrefix` is money

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackOnboarder.scala (hunk 17 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackOnboarder.scala:169]: seems slightly more risky than it's worth. Some paths have different `trackingParams` within the same scope e.g. subactions.

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala (hunk 9 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala:435]: started at this... stopped at it.

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala [hunk 19]
```diff
@@ -41,11 +46,27 @@ object Path {
...
-  def alreadyEncoded(value: String): EncodedPath = new EncodedPath(URLDecoder.decode(value, UTF8))
+  def encode(value: String): EncodedPath = {
+    val (path, query) = splitQuery(value)
+    new EncodedPath(path, query)
+  }
+
+  def alreadyEncoded(value: String): EncodedPath = {
+    val (path, query) = splitQuery(URLDecoder.decode(value, UTF8))
+    new EncodedPath(path, query)
+  }
```
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala:66]: @ryanpbrewster I would appreciate your feedback on these updates to `Path`.

File: kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala [hunk 20]
```diff
@@ -173,4 +174,38 @@ class URITest extends Specification {
...
+      URI.parse(queriedUrl) === URI.parse(queriedPath.absolute)
+      queriedPath.queryString === simpleQuery
+
+      val complexQuery = "&b=c&c=d&d=e"
+      val complexQueriedUrl = queriedUrl + complexQuery
+      val complexQueriedPath = queriedPath.withQuery(Query.parse(complexQuery))
+      complexQueriedUrl === (queriedPath + complexQuery).absolute
+      complexQueriedUrl === complexQueriedPath.absolute
+      complexQueriedPath.queryString === simpleQuery + complexQuery
+    }
```
Comment (camhashemi) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: @ryanpbrewster tests for those updates

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala [hunk 21]
```diff
@@ -41,11 +46,27 @@ object Path {
 
   def base: String = "https://www.kifi.com/"
 
-  def apply(value: String): Path = new Path(value)
+  private def splitQuery(str: String): (String, Query) = str.split('?').toList match {
+    case relative :: params :: Nil => (relative, Query.parse(params))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: This is handled by the `relative :: params =>` case, if you switch from `reduce` to `foldLeft(Query.empty)`

File: kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala (hunk 20 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: What should happen if I try to do something like `Path("") + "?a=b" + "?c=d"? Exceptions?

File: kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala [hunk 22]
```diff
@@ -231,9 +236,23 @@ class SlackPushingActor @Inject() (
...
+                case _: PushItem.Digest => NotificationCategory.NonUser.LIBRARY_DIGEST
+                case PushItem.KeepToPush(k, ktl) =>
+                  contextBuilder += ("keepId", k.id.get.id)
+                  contextBuilder += ("libraryId", ktl.libraryId.id)
+                  NotificationCategory.NonUser.NEW_KEEP
+                case PushItem.MessageToPush(k, _) =>
+                  contextBuilder += ("threadId", k.id.get.id)
+                  NotificationCategory.NonUser.NEW_COMMENT
+              }
+              slackAnalytics.trackNotificationSent(integration.slackTeamId, integration.slackChannelId, integration.slackChannelName, category, contextBuilder.build)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala:256]: I think we should track a sent notif after we have confirmed it went through (i.e., this should go in an `andThen` or a `map` on the future

Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/slack/SlackPushingActor.scala:256]: good call, thanks

File: kifi-backend/modules/common/app/com/keepit/common/path/Path.scala (hunk 21 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/common/path/Path.scala]: :+1: huge improvement

File: kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala (hunk 20 above)
Comment (camhashemi) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: exceptions would be pretty bad haha, it should append the query parameters and normalize within reason. I added a test to show it.

File: kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala [hunk 23]
```diff
@@ -173,4 +174,47 @@ class URITest extends Specification {
...
+      complexQueriedUrlString === complexQueriedPath.absolute
+      complexQueriedPath.queryString === simpleQueryString + complexQueryString
+
+      val complexQueryString2 = "?e=f&f=g"
+      val complexQueriedUrlString2 = complexQueriedUrlString + "&" + complexQueryString2.stripPrefix("?")
+      val complexQueriedPath2 = complexQueriedPath.withQuery(Query.parse(complexQueryString2))
+      complexQueriedUrlString2 === (complexQueriedPath + complexQueryString2).absolute
+      complexQueriedUrlString2 === complexQueriedPath2.absolute
+      complexQueriedPath2.queryString === simpleQueryString + complexQueryString + "&" + complexQueryString2.stripPrefix("?")
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: It might be instructive to have simple "input to output" tests, like

```
    // path segments, with and without leading '/'
    (Path("") + "kifi").absolute === "http://www.kifi.com/kifi"
    (Path("") + "/kifi").absolute === "http://www.kifi.com/kifi"

    // path segment addition must have leading '/' if desired
    (Path("kifi") + "general").absolute === "http://www.kifi.com/kifigeneral"
    (Path("kifi") + "/general").absolute === "http://www.kifi.com/kifi/general"

    // query parameters should be prefixed with a '?' or '&', and automatically do The Right Thing (TM)
    (Path("kifi/general") + "?a=b").absolute === "http://www.kifi.com/kifi/general?a=b"
    (Path("kifi/general") + "&a=b").absolute === "http://www.kifi.com/kifi/general?a=b"
    (Path("kifi/general?a=b") + "&c=d").absolute === "http://www.kifi.com/kifi/general?a=b&c=d"
    (Path("kifi/general?a=b") + "?c=d").absolute === "http://www.kifi.com/kifi/general?a=b&c=d"

    // multiple query params can be added at once if you manually inject the '&' between them
    (Path("kifi/general") + "?a=b&c=d").absolute === "http://www.kifi.com/kifi/general?a=b&c=d"
    (Path("kifi/general?a=b") + "&c=d&e=f").absolute === "http://www.kifi.com/kifi/general?a=b&c=d&e=f"
    (Path("kifi/general?a=b") + "?c=d&e=f").absolute === "http://www.kifi.com/kifi/general?a=b&c=d&e=f"
```

Comment (camhashemi) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: minus the `http: ->  https:` refactor, all these tests pass. you're a beast.

Comment (ryanpbrewster) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: :boom: :metal:

Comment (camhashemi) [inline kifi-backend/modules/common/test/com/keepit/common/net/URITest.scala:196]: I'm going to add to them, it's nice documentation for all the edge cases and implicit behavior. Thanks a lot.
